File: hangman_lua/RandomWords.lua
local Set = require 'Set'

if not arg or #arg < 1 then
  io.stderr:write("A number of random words is expected but missing.\n")
  os.exit(false, true)
end

local DICT_FILE = os.getenv("hangman_dict") or "words.txt"

local dict = {}
local f = io.open(DICT_FILE, "rb")
if not f then
  io.stderr:write(string.format("Cannot open dictionary file '%s' for reading!\n", DICT_FILE))
  os.exit(false, true)
end

for line in f:lines() do
  if #line > 0 then
    table.insert(dict, line:upper())
  end
end

local count = tonumber(arg[1])
if count < 1 or count > #dict then
  io.stderr:write(string.format("%s is not a number, or its value is out of [1, %d]\n", arg[1], #dict))
  os.exit(false, true)
end

local out = Set()
local size = #dict
math.randomseed(os.time())
local i = count
while i >= 1 do
  local idx = (math.random(size) * 10 + math.random(size) + math.floor(math.random(size) / (math.random(size) + 1))) % size + 1
  if out:add(idx) then
    io.stdout:write(dict[idx]:upper(), "\n")
    i = i - 1
  end
end

File: trunk/hangman_cpp/RandomWords.cpp
#include <string>
#include <vector>
#include <set>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <assert.h>
#include "Utility.h"

using namespace std;

int main(int argc, char * argv[]) {
  if (argc < 2) {
    fprintf(stderr, "usage:\n%s <number-of-words>\n", argv[0]);
    return -1;
  }

  const char * file = getenv("hangman_dict");
  if (!file)
    file = "words.txt";

  FILE * f = fopen(file, "rb");
  if (!f) {
    fprintf(stderr, "Cannot open dictionary file '%s' for reading!\n", file);
    return -1;
  }

  // Read in dictionary file
  vector<string> dict;
  char wordbuf[1024];
  while (!feof(f)) {
    string word = readWord(f);
    if (!word.empty()) {
      dict.push_back(word);
    }
  }
  fclose(f);

  int count = atoi(argv[1]);
  if (count < 1 || count > dict.size()) {
    fprintf(stderr, "%s is not a number, or its value is out of [1, %d]\n", argv[1], dict.size());
    return -1;
  }

  // Produce random words in dictionary
  unsigned int size = dict.size();
  set<unsigned int> included;
  srand(time(NULL));

  while (count > 0) {
    unsigned int r = (unsigned int)(rand() * 10 + rand() + (rand() / (rand() + 1))) % size;
    if (included.insert(r).second) {
      fprintf(stdout, "%s\n", dict[r].c_str());
      count--;
    }
  }

  return 0;
}
